NullPointerException in org.ocpsoft.urlbuilder.AddressBuilder#queryDecoded

Method throws NullPointerException with given input (line 204)
CharSequence name = null
Object... values = Object[0]

Method called from org.ocpsoft.rewrite.servlet.config.PathAndQuery#evaluateHttp (line 133)

InboundRewrite [POST url=http://localhost:8080/rest/authenticationServices/validateUserCredentials, flow=UN_HANDLED, dispatchResource=null]

```
    <dependency>
        <groupId>org.ocpsoft.rewrite</groupId>
        <artifactId>rewrite-servlet</artifactId>
        <version>3.4.1.Final</version>
    </dependency>
    <dependency>
        <groupId>org.ocpsoft.rewrite</groupId>
        <artifactId>rewrite-integration-cdi</artifactId>
        <version>3.4.1.Final</version>
    </dependency>
```

Some more details, sorry for not mentioning. Upgraded from 2.0.12.Final to 3.4.1.Final, with 2.0.12 there was no problem with the code.

Here the full stack trace
```
Caused by: java.lang.NullPointerException
        at org.ocpsoft.urlbuilder.AddressBuilder.queryDecoded(AddressBuilder.java:204) [rewrite-servlet-3.4.1.Final.jar:3.4.1.Final]
        at org.ocpsoft.urlbuilder.AddressBuilderPath.queryDecoded(AddressBuilderPath.java:58) [rewrite-servlet-3.4.1.Final.jar:3.4.1.Final]
        at org.ocpsoft.rewrite.servlet.config.PathAndQuery.evaluateHttp(PathAndQuery.java:134) [rewrite-servlet-3.4.1.Final.jar:3.4.1.Final]
        at org.ocpsoft.rewrite.servlet.config.HttpCondition.evaluate(HttpCondition.java:44) [rewrite-servlet-3.4.1.Final.jar:3.4.1.Final]
        at org.ocpsoft.rewrite.config.RuleBuilder.evaluate(RuleBuilder.java:156) [rewrite-servlet-3.4.1.Final.jar:3.4.1.Final]
        at org.ocpsoft.rewrite.servlet.impl.DefaultHttpRewriteProvider.rewriteInbound(DefaultHttpRewriteProvider.java:205) [rewrite-servlet-3.4.1.Final.jar:3.4.1.Final]
```

Were able to fix the issue by replacing PathAndQuery.matches with Path.matches as our rewrite rules handles request path values only.

--- a/addressbuilder/src/main/java/org/ocpsoft/urlbuilder/AddressBuilder.java
+++ b/addressbuilder/src/main/java/org/ocpsoft/urlbuilder/AddressBuilder.java
@@ -182,7 +182,10 @@ public class AddressBuilder
     */
    AddressBuilderQuery query(CharSequence name, Object... values)
    {
-      this.queries.put(name.toString(), Parameter.create(name.toString(), values));
+      if (name != null && values != null)
+      {
+         this.queries.put(name.toString(), Parameter.create(name.toString(), values));
+      }
       return new AddressBuilderQuery(this);
    }
 
@@ -191,7 +194,7 @@ public class AddressBuilder
     */
    AddressBuilderQuery queryDecoded(CharSequence name, Object... values)
    {
-      if (values != null)
+      if (name != null && values != null)
       {
          List<Object> encodedValues = new ArrayList<Object>(values.length);
          for (Object value : values)
@@ -211,7 +214,7 @@ public class AddressBuilder
     */
    AddressBuilderQuery queryEncoded(CharSequence name, Object... values)
    {
-      if (values != null)
+      if (name != null && values != null)
       {
          List<Object> encodedValues = new ArrayList<Object>(values.length);
          for (Object value : values)
@@ -323,7 +326,7 @@ public class AddressBuilder
     */
    void setDecoded(CharSequence name, Object... values)
    {
-      if (values != null)
+      if (name != null && values != null)
       {
          List<Object> encodedValues = new ArrayList<Object>(values.length);
          for (Object value : values)
@@ -342,7 +345,7 @@ public class AddressBuilder
     */
    void setEncoded(CharSequence name, Object... values)
    {
-      if (values != null)
+      if (name != null && values != null)
       {
          List<Object> encodedValues = new ArrayList<Object>(values.length);
          for (Object value : values)

--- a/addressbuilder/src/test/java/org/ocpsoft/urlbuilder/AddressBuilderTest.java
+++ b/addressbuilder/src/test/java/org/ocpsoft/urlbuilder/AddressBuilderTest.java
@@ -314,4 +314,33 @@ public class AddressBuilderTest
       assertEquals("http://localhost/somepath/%7Bsomething%7D", address.toString());
    }
 
+   @Test
+   public void testQueryWithNull() {
+      assertEquals("",
+              AddressBuilder.begin().query(null).build().toString());
+   }
+
+   @Test
+   public void testQueryDecodedWithNull() {
+      assertEquals("",
+              AddressBuilder.begin().queryDecoded(null).build().toString());
+   }
+
+   @Test
+   public void testQueryEncodedWithNull() {
+      assertEquals("",
+              AddressBuilder.begin().queryEncoded(null).build().toString());
+   }
+
+   @Test
+   public void testSetEncodedWithNull() {
+      assertEquals("//localhost",
+              AddressBuilder.begin().domain("localhost").setEncoded(null).build().toString());
+   }
+
+   @Test
+   public void testSetDecodedWithNull() {
+      assertEquals("//localhost",
+              AddressBuilder.begin().domain("localhost").setDecoded(null).build().toString());
+   }
 }

--- a/config-servlet/src/main/java/org/ocpsoft/rewrite/servlet/config/PathAndQuery.java
+++ b/config-servlet/src/main/java/org/ocpsoft/rewrite/servlet/config/PathAndQuery.java
@@ -130,8 +130,10 @@ public abstract class PathAndQuery extends HttpCondition implements Parameterize
             return false;
       }
       else
-         url = AddressBuilder.begin().pathDecoded(event.getInboundAddress().getPath())
-                  .queryDecoded(event.getInboundAddress().getQuery()).build().toString();
+         url = AddressBuilder.begin()
+                 .pathDecoded(event.getInboundAddress().getPath())
+                 .queryLiteral(event.getInboundAddress().getQuery())
+                 .build().toString();
 
       String contextPath = event.getContextPath();
       if (!contextPath.equals("/") && url.startsWith(contextPath))

--- a/config-servlet/src/test/java/org/ocpsoft/rewrite/servlet/config/PathAndQueryProvider.java
+++ b/config-servlet/src/test/java/org/ocpsoft/rewrite/servlet/config/PathAndQueryProvider.java
@@ -1,0 +1,43 @@
+/*
+ * Copyright 2016 <a href="mailto:lincolnbaxter@gmail.com">Lincoln Baxter, III</a>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ocpsoft.rewrite.servlet.config;
+
+import javax.servlet.ServletContext;
+
+import org.ocpsoft.rewrite.config.Configuration;
+import org.ocpsoft.rewrite.config.ConfigurationBuilder;
+
+/**
+ * Configuration provider for {@link PathAndQueryTest}.
+ */
+public class PathAndQueryProvider extends HttpConfigurationProvider
+{
+
+   @Override
+   public int priority()
+   {
+      return 0;
+   }
+
+   @Override
+   public Configuration getConfiguration(final ServletContext context)
+   {
+      return ConfigurationBuilder.begin()
+              .addRule()
+              .when(PathAndQuery.matches("/pathAndQuery/{tail}"))
+              .perform(Redirect.temporary("/resultingPath/{tail}"));
+   }
+}

--- a/config-servlet/src/test/java/org/ocpsoft/rewrite/servlet/config/PathAndQueryTest.java
+++ b/config-servlet/src/test/java/org/ocpsoft/rewrite/servlet/config/PathAndQueryTest.java
@@ -1,0 +1,64 @@
+/*
+ * Copyright 2016 <a href="mailto:lincolnbaxter@gmail.com">Lincoln Baxter, III</a>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ocpsoft.rewrite.servlet.config;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.arquillian.junit.Arquillian;
+import org.jboss.arquillian.test.api.ArquillianResource;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.ocpsoft.rewrite.config.ConfigurationProvider;
+import org.ocpsoft.rewrite.test.RewriteTest;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.htmlunit.HtmlUnitDriver;
+
+/**
+ * @see https://github.com/ocpsoft/rewrite/issues/243
+ */
+@RunWith(Arquillian.class)
+public class PathAndQueryTest extends RewriteTest
+{
+
+   @Deployment(testable = false)
+   public static WebArchive getDeployment()
+   {
+      return RewriteTest.getDeployment()
+              .addClass(PathAndQueryProvider.class)
+              .addAsServiceProvider(ConfigurationProvider.class, PathAndQueryProvider.class);
+   }
+
+   @ArquillianResource
+   private java.net.URL baseUrl;
+
+   @Test
+   public void testWithQuery() throws Exception
+   {
+      WebDriver driver = new HtmlUnitDriver();
+      driver.get(baseUrl.toString() + "pathAndQuery/pathWithQuery?p1=foo&p2=bar");
+      assertThat(driver.getCurrentUrl()).endsWith("/resultingPath/pathWithQuery%3Fp1=foo&p2=bar");
+   }
+
+   @Test
+   public void testWithoutQuery() throws Exception
+   {
+      WebDriver driver = new HtmlUnitDriver();
+      driver.get(baseUrl.toString() + "pathAndQuery/pathWithoutQuery");
+      assertThat(driver.getCurrentUrl()).endsWith("/resultingPath/pathWithoutQuery");
+   }
+}

